dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/relation/RelationModelManager.java

```diff
@@ -0,0 +1,596 @@
+    }
+
+    private void extractRelations(List<String> relationNames) {
+        Set<Relation> relsToExtract = new HashSet<>();
+        for (String name : relationNames) {
+            Relation r = getRelationWithName(name);
+            if (r == null) {
+                logger.warn("Relation with the name {} does not exist", name);
+                continue;
+            }
+            relsToExtract.add(r);
+            createModel(r, name);
+        }
+
+        Set<RelationGraph> relGraphs = new HashSet<>();
+        DependencyGraph<Relation> dependencyGraph = DependencyGraph.from(relsToExtract);
+
+        for (Set<DependencyGraph<Relation>.Node> component : dependencyGraph.getSCCs()) {
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = new RecursiveGraph();
+                    recGraph.setName(r.getNameOrTerm() + "_rec");
+                    relGraphs.add(recGraph);
+                    relGraphCache.put(r, recGraph);
+                }
+            }
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                RelationGraph rg = createGraph(r);
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = (RecursiveGraph) relGraphCache.get(r);
+                    recGraph.setConcreteGraph(rg);
+                } else {
+                    relGraphs.add(rg);
+                }
+            }
+        }
+
+        Set<CAATPredicate> predicates = new HashSet<>(relGraphs);
+        PredicateHierarchy hierarchy = new PredicateHierarchy(predicates);
+        hierarchy.initializeToDomain(domain);
+
+        // Populate graphs of base relations.
+        for (CAATPredicate basePred : hierarchy.getBasePredicates()) {
+            Relation r = relGraphCache.inverse().get((RelationGraph) basePred);
+            try {
+                r.getDefinition().accept(graphPopulator);
+            } catch (UnsupportedOperationException e) {
+                graphPopulator.populateDynamicDefaultGraph(r);
+            }
+        }
+
+        // Do the computation.
+        hierarchy.populate();
+
+        for (CAATPredicate pred : hierarchy.getPredicateList()) {
+            Relation r = relGraphCache.inverse().get((RelationGraph) pred);
+            if (relModelCache.containsKey(r)) {
+                RelationModel rm = relModelCache.get(r);
+                ((RelationGraph) pred).edgeStream()
+                                      .forEach(e -> rm.addEdgeModel(getOrCreateEdgeModel(e)));
+           }
+        }
+
+        for (Relation r : relsToExtract) {
+            RelationModel rm;
+            if (r.getDefinition().getClass() == Coherence.class) {
+                rm = constructCoherenceModel(r);
+            } else {
+                rm = relModelCache.get(r);
+            }
```

Comment:
This would probably read better as
```
RelationModel rm = r.getDefinition().getClass() == Coherence.class ? constructCoherenceModel(r) : relModelCache.get(r);
```